### DISCUSSION TITLE
MySQL Query Builder: critical query for large mysql instances

### DISCUSSION THREAD

**Discussion by irgendwer112** [grafana/grafana#37002] [Data Sources]
The new query builder uses information_schema queries for supporting the editor.

But: this query is killing large MySQL instances with some 100.000 tables.

<<<
// query that returns first table found that has a timestamp(tz) column and a float column
const query = `
SELECT
table_name as table_name,
( SELECT
column_name as column_name
FROM information_schema.columns c
WHERE
c.table_schema = t.table_schema ...
>>>

Result is a very long running locking query with "checking permissions" state. Mysql information_schema still has some problems with large installations, so most utilities allow to disable usage of information_schema.

This would be a nice config option in grafana too, even by removing support for the query editor or allow a more granular configuration for some databases with usage of "SHOW TABLES FROM %database" instead of information_schema.

**Comment by marefr** [comment]
Too me `information_schema` should be used since it's an ANSI standard and `SHOW TABLES` seems like a MySQL pre-5.0 solution?

Would it help you to add a filter for the actual database, e.g. `table_schema = DATABASE()` or similar?

**Comment by irgendwer112** [comment]
Yes: limit table_schema to current database should improve the query.

**Comment by hackalot805** [comment]
I was bit by this yesterday.  It caused our replication slave to eat up swap pages and slowed the replication queries to crawl.  Any plans to optimize this soon?  Or a setting to turn off what ever function is needing to run this query might be nice.

**Comment by marefr** [comment]
The fix of adding `table_schema = DATABASE()` needs to be implemented in this query:
https://github.com/grafana/grafana/blob/6fbca90269a0fe17f4e5ab7dac608b71ddd94db6/public/app/plugins/datasource/mysql/meta_query.ts#L29-L74

Anyone interested in testing if query works better with schema constraint and/or contributing a fix?

**Comment by gabor** [comment]
the current version of the mentioned codebase has quite changed, and right now it contains the `WHERE table_schema = ${database}` part ( https://github.com/grafana/grafana/blob/c755ceae314071e1dd30d62c329d6f2590f45514/public/app/plugins/datasource/mysql/mySqlMetaQuery.ts#L5C60-L5C92 ),
so i assume we can consider the improvement implemented.

if this solution is not sufficient, please open a new issue with the problem,and why this solution is not sufficient. thanks!
